{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "soi = ['kluyveromyces_marxianus', 'komagataella_phaffii', 'yarrowia_lipolytica']\n",
    "\n",
    "df = pd.read_csv('data/input/fourdbs_hi_input_species.csv')\n",
    "# df[~df['species_name'].isin(soi)].reset_index(drop=True).to_csv('data/input/fourdbs_hi_without_3soi.csv', index=False)\n",
    "# df[df['species_name'].isin(soi)].reset_index(drop=True).to_csv('data/input/3soi.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run allegro on without_3soi from above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "from src.utils.guide_finder import GuideFinder\n",
    "\n",
    "\n",
    "def find_all_matches(list1, list2):\n",
    "    matches = list()\n",
    "\n",
    "    for index, item in enumerate(list2):\n",
    "        if item in list1:\n",
    "            matches.append(index)\n",
    "        \n",
    "    return matches\n",
    "\n",
    "def hamming_distance(str1: str, str2: str, length: int) -> int:\n",
    "    \"\"\"\n",
    "    Calculates the Hamming distance between two strings up to a specified length.\n",
    "    \"\"\"\n",
    "    if len(str1) < length or len(str2) < length:\n",
    "        raise ValueError('Strings must have a length of at least {n}'.format(n=length))\n",
    "\n",
    "    return sum(ch1 != ch2 for ch1, ch2 in zip(str1[:length], str2[:length]))\n",
    "\n",
    "\n",
    "def get_record_metadata(record):\n",
    "    # This gene's own name -- Usually N/A for unannotated CDS files or chromosomes/scaffolds\n",
    "    gene_match = re.search(gene_regex, record.description)\n",
    "    gene_name = gene_match.group(1) if gene_match is not None else 'N/A'\n",
    "\n",
    "    # This gene's own protein id e.g., XP_022674739.1\n",
    "    protein_id_match = re.search(protein_id_regex, record.description)\n",
    "    protein_id = protein_id_match.group(1) if protein_id_match is not None else 'N/A'\n",
    "\n",
    "    # For example, [orthologous_to_ref_protein=XP_022674739.1], extracts XP_022674739.1\n",
    "    ortho_prot_to_match = re.search(orthologous_protein_regex, record.description)\n",
    "    ortho_prot_id = ortho_prot_to_match.group(1) if ortho_prot_to_match is not None else 'N/A'\n",
    "\n",
    "    # For example, [orthologous_to_gene=HIS7], extracts HIS7\n",
    "    ortho_gene_to_match = re.search(orthologous_name_regex, record.description)\n",
    "    ortho_gene_name = ortho_gene_to_match.group(1) if ortho_gene_to_match is not None else 'N/A'\n",
    "\n",
    "    return gene_name, protein_id, ortho_prot_id, ortho_gene_name\n",
    "\n",
    "\n",
    "gene_regex = r'\\[gene=(.*?)\\]'\n",
    "tag_regex = r'\\[locus_tag=(.*?)\\]'\n",
    "protein_id_regex = r'\\[protein_id=(.*?)\\]'\n",
    "reference_species_regex = r'\\[ref_species=(.*?)\\]'\n",
    "orthologous_name_regex = r'\\[orthologous_to_gene=(.*?)\\]'\n",
    "orthologous_protein_regex = r'\\[orthologous_to_ref_protein=(.*?)\\]'\n",
    "\n",
    "new_exp_name = 'all_without_threeSOI_e1'\n",
    "\n",
    "split_out = pd.read_csv(f'data/output/{new_exp_name}/{new_exp_name}.csv')\n",
    "\n",
    "library = split_out.sequence.unique()\n",
    "\n",
    "test_df = pd.read_csv(f'data/input/3soi.csv')[['species_name', 'cds_file_name']]\n",
    "\n",
    "df_species_name_list = list()\n",
    "df_covered_list = list()\n",
    "df_strands = list()\n",
    "df_mismatch = list()\n",
    "df_locations = list()\n",
    "df_gene_names = list()\n",
    "df_protein_ids = list()\n",
    "df_ortho_prot_ids = list()\n",
    "df_ortho_gene_names = list()\n",
    "df_lib_partial_match = list()\n",
    "\n",
    "gf = GuideFinder()\n",
    "base_path = 'data/input/cds/orthogroups/'\n",
    "\n",
    "for label, row in test_df.iterrows():\n",
    "    species_name = row['species_name']\n",
    "\n",
    "    species_is_covered_n_times = 0\n",
    "    exact_matching_guides = list()\n",
    "\n",
    "    records_path = base_path + row['cds_file_name']\n",
    "    records = list(SeqIO.parse(open(records_path), 'fasta'))\n",
    "\n",
    "    for record in records:\n",
    "        gene_is_covered_n_times = 0\n",
    "        \n",
    "        gene_name, protein_id, ortho_prot_id, ortho_gene_name = get_record_metadata(record)\n",
    "\n",
    "        (guides_list,\n",
    "        _guides_context_list,\n",
    "        strands_list,\n",
    "        locations_list) = gf.identify_guides_and_indicate_strand(\n",
    "            pam='NGG',\n",
    "            sequence=str(record.seq).upper(),\n",
    "            protospacer_length=20,\n",
    "            context_toward_five_prime=0,\n",
    "            context_toward_three_prime=0,\n",
    "            filter_repetitive=False\n",
    "        )\n",
    "\n",
    "        # Attempt to find an exact match.\n",
    "        matches = find_all_matches(library, guides_list)\n",
    "\n",
    "        # No exact matches found for this gene.\n",
    "        if len(matches) == 0:\n",
    "\n",
    "            # Try partial matches for this genes.\n",
    "            mm_allowed = 2\n",
    "            req_match_len = 10\n",
    "\n",
    "            # needs to prioritize 1 mismatch first, then 2 \n",
    "\n",
    "            for idx, test_guide in enumerate(guides_list):\n",
    "                for lib_guide in library:\n",
    "\n",
    "                    distance_after_seed = hamming_distance(lib_guide, test_guide, req_match_len)\n",
    "                    \n",
    "                    if (lib_guide[req_match_len:] == test_guide[req_match_len:]) and \\\n",
    "                     (distance_after_seed <= mm_allowed) and \\\n",
    "                     (distance_after_seed > 0):\n",
    "                     \n",
    "                        df_species_name_list.append(species_name)\n",
    "                        df_covered_list.append(test_guide)\n",
    "                        df_locations.append(locations_list[idx])\n",
    "                        df_strands.append(strands_list[idx])\n",
    "                        df_gene_names.append(gene_name)\n",
    "                        df_protein_ids.append(protein_id)\n",
    "                        df_ortho_prot_ids.append(ortho_prot_id)\n",
    "                        df_ortho_gene_names.append(ortho_gene_name)\n",
    "                        df_mismatch.append(distance_after_seed)\n",
    "                        df_lib_partial_match.append(lib_guide)\n",
    "                        gene_is_covered_n_times += 1\n",
    "                        species_is_covered_n_times += 1\n",
    "\n",
    "        else:\n",
    "            for match in matches:\n",
    "                df_species_name_list.append(species_name)\n",
    "                df_covered_list.append(guides_list[match])\n",
    "                df_locations.append(locations_list[match])\n",
    "                df_strands.append(strands_list[match])\n",
    "                df_gene_names.append(gene_name)\n",
    "                df_protein_ids.append(protein_id)\n",
    "                df_ortho_prot_ids.append(ortho_prot_id)\n",
    "                df_ortho_gene_names.append(ortho_gene_name)\n",
    "                df_lib_partial_match.append('N/A')\n",
    "                df_mismatch.append('N/A')\n",
    "                gene_is_covered_n_times += 1\n",
    "                species_is_covered_n_times += 1\n",
    "                exact_matching_guides.append(match)\n",
    "\n",
    "        # Did all we could. Gene is still uncovered:\n",
    "        if gene_is_covered_n_times == 0:\n",
    "            df_species_name_list.append(species_name)\n",
    "            df_covered_list.append('Uncovered gene')\n",
    "            df_locations.append('Uncovered gene')\n",
    "            df_strands.append('Uncovered gene')\n",
    "            df_gene_names.append(gene_name)\n",
    "            df_protein_ids.append(protein_id)\n",
    "            df_ortho_prot_ids.append(ortho_prot_id)\n",
    "            df_ortho_gene_names.append(ortho_gene_name)\n",
    "            df_lib_partial_match.append('N/A')\n",
    "            df_mismatch.append('N/A')\n",
    "    \n",
    "\n",
    "    # No exact or partial matches anywhere at all in this species.\n",
    "    if species_is_covered_n_times == 0:\n",
    "        df_species_name_list.append(species_name)\n",
    "        df_covered_list.append('Not covered')\n",
    "        df_strands.append('Not covered')\n",
    "        df_locations.append('Not covered')\n",
    "        df_gene_names.append('Not covered')\n",
    "        df_protein_ids.append('Not covered')\n",
    "        df_ortho_prot_ids.append('Not covered')\n",
    "        df_ortho_gene_names.append('Not covered')\n",
    "        df_mismatch.append('Not covered')\n",
    "        df_lib_partial_match.append('Not covered')\n",
    "\n",
    "test_df = pd.DataFrame.from_dict({\n",
    "    'species': df_species_name_list,\n",
    "    'covered_by': df_covered_list,\n",
    "    'partial_match_in_library': df_lib_partial_match,\n",
    "    'mismatch': df_mismatch,\n",
    "    'strand': df_strands,\n",
    "    'location': df_locations,\n",
    "    'gene_name': df_gene_names,\n",
    "    'protein_id': df_protein_ids,\n",
    "    'ortho_prot_id': df_ortho_prot_ids,\n",
    "    'ortho_gene_name': df_ortho_gene_names,\n",
    "})\n",
    "\n",
    "test_df.to_csv(f'data/output/test_{new_exp_name}_results.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>covered_by</th>\n",
       "      <th>partial_match_in_library</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>strand</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ortho_prot_id</th>\n",
       "      <th>ortho_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CTCGACATTGTGCAGCTGCA</td>\n",
       "      <td>CTGGACATCGTGCAGCTGCA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>F</td>\n",
       "      <td>854</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_500614.1</td>\n",
       "      <td>NP_010290.3</td>\n",
       "      <td>TRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_500977.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_500977.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>ATCTCGCTTGGAGGAACCAT</td>\n",
       "      <td>ATCTCCCTTGGAGGAACCAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F</td>\n",
       "      <td>236</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_501094.1</td>\n",
       "      <td>NP_010851.1</td>\n",
       "      <td>CAN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CTTGGAGGAACCATTGGTAC</td>\n",
       "      <td>CTTGGTGGTACCATTGGTAC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>F</td>\n",
       "      <td>242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_501094.1</td>\n",
       "      <td>NP_010851.1</td>\n",
       "      <td>CAN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>ATTGGTACCGGTCTCTTCAT</td>\n",
       "      <td>ATCGGCACCGGTCTCTTCAT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>F</td>\n",
       "      <td>254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_501094.1</td>\n",
       "      <td>NP_010851.1</td>\n",
       "      <td>CAN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>ATTGGTACCGGTCTCTTCAT</td>\n",
       "      <td>ATCGGTACCGGTCTCTTCAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>F</td>\n",
       "      <td>254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_501094.1</td>\n",
       "      <td>NP_010851.1</td>\n",
       "      <td>CAN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>ACCAAGTGGATCGGAGGCCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_503263.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>GATGACCAGGTGAAGATTCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>2177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_503627.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>GCTCCGGTGCACATATCACA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RC</td>\n",
       "      <td>204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_505753.2</td>\n",
       "      <td>NP_015387.1</td>\n",
       "      <td>FCY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>AAACAATGGAGCTTCATTAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>76</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002489365.1</td>\n",
       "      <td>NP_010290.3</td>\n",
       "      <td>TRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>CATTTGCAGATGATTGCCAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>248</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002489665.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>GTAAACATATCTCTTTGAAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RC</td>\n",
       "      <td>2775</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002490264.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>TTTAGATACTGGAGAAACCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>713</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002490740.1</td>\n",
       "      <td>NP_010851.1</td>\n",
       "      <td>CAN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>CAGTTCTTGTATGTCTTTCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RC</td>\n",
       "      <td>249</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002490927.1</td>\n",
       "      <td>NP_015387.1</td>\n",
       "      <td>FCY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>GACAGAAAGTTTGCTGATAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002492605.1</td>\n",
       "      <td>NP_010893.3</td>\n",
       "      <td>URA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>GTTGACAACACCTTTGGTGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002493751.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>komagataella_phaffii</td>\n",
       "      <td>TTCTCCAACTTGGCTAACGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>1133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_002493751.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GAGTTAGTCGGTATCACTGC</td>\n",
       "      <td>GAATTGGTCGGTATCACTGC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>F</td>\n",
       "      <td>884</td>\n",
       "      <td>CAN1</td>\n",
       "      <td>XP_022673643.1</td>\n",
       "      <td>NP_010851.1</td>\n",
       "      <td>CAN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>CACTTGCAGATGATCGCCAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>275</td>\n",
       "      <td>GAP1</td>\n",
       "      <td>XP_022673914.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>ATTGGTGGTGCCATTGGTAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>293</td>\n",
       "      <td>GAP1</td>\n",
       "      <td>XP_022673914.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>TGGTACTTACAACCAATTCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>352</td>\n",
       "      <td>MET17</td>\n",
       "      <td>XP_022674188.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>TCTGCTACCAAGTGGATTGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>632</td>\n",
       "      <td>MET17</td>\n",
       "      <td>XP_022674188.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>ACCAAGTGGATTGGTGGTCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>638</td>\n",
       "      <td>MET17</td>\n",
       "      <td>XP_022674188.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GACAGGAAGTTTGCCGACAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>290</td>\n",
       "      <td>URA3</td>\n",
       "      <td>XP_022674242.1</td>\n",
       "      <td>NP_010893.3</td>\n",
       "      <td>URA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GGTGAGAGATACAGAAAGGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>761</td>\n",
       "      <td>URA3</td>\n",
       "      <td>XP_022674242.1</td>\n",
       "      <td>NP_010893.3</td>\n",
       "      <td>URA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>Uncovered gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncovered gene</td>\n",
       "      <td>Uncovered gene</td>\n",
       "      <td>TRP1</td>\n",
       "      <td>XP_022677214.1</td>\n",
       "      <td>NP_010290.3</td>\n",
       "      <td>TRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GTGAACATATCTCTTTGGAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RC</td>\n",
       "      <td>2769</td>\n",
       "      <td>LYS2</td>\n",
       "      <td>XP_022678348.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>TCTCCATGTGATATGTGCAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>266</td>\n",
       "      <td>FCY1</td>\n",
       "      <td>XP_022678383.1</td>\n",
       "      <td>NP_015387.1</td>\n",
       "      <td>FCY1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    species            covered_by partial_match_in_library  \\\n",
       "0       yarrowia_lipolytica  CTCGACATTGTGCAGCTGCA     CTGGACATCGTGCAGCTGCA   \n",
       "1       yarrowia_lipolytica  CATCTCCAGATGATTGCCAT                      NaN   \n",
       "2       yarrowia_lipolytica  CTCCAGATGATTGCCATCGG                      NaN   \n",
       "3       yarrowia_lipolytica  ATCTCGCTTGGAGGAACCAT     ATCTCCCTTGGAGGAACCAT   \n",
       "4       yarrowia_lipolytica  CTTGGAGGAACCATTGGTAC     CTTGGTGGTACCATTGGTAC   \n",
       "5       yarrowia_lipolytica  ATTGGTACCGGTCTCTTCAT     ATCGGCACCGGTCTCTTCAT   \n",
       "6       yarrowia_lipolytica  ATTGGTACCGGTCTCTTCAT     ATCGGTACCGGTCTCTTCAT   \n",
       "7       yarrowia_lipolytica  ACCAAGTGGATCGGAGGCCA                      NaN   \n",
       "8       yarrowia_lipolytica  GATGACCAGGTGAAGATTCG                      NaN   \n",
       "9       yarrowia_lipolytica  GCTCCGGTGCACATATCACA                      NaN   \n",
       "10     komagataella_phaffii  AAACAATGGAGCTTCATTAG                      NaN   \n",
       "11     komagataella_phaffii  CATTTGCAGATGATTGCCAT                      NaN   \n",
       "12     komagataella_phaffii  GTAAACATATCTCTTTGAAT                      NaN   \n",
       "13     komagataella_phaffii  TTTAGATACTGGAGAAACCC                      NaN   \n",
       "14     komagataella_phaffii  CAGTTCTTGTATGTCTTTCC                      NaN   \n",
       "15     komagataella_phaffii  GACAGAAAGTTTGCTGATAT                      NaN   \n",
       "16     komagataella_phaffii  GTTGACAACACCTTTGGTGC                      NaN   \n",
       "17     komagataella_phaffii  TTCTCCAACTTGGCTAACGT                      NaN   \n",
       "18  kluyveromyces_marxianus  GAGTTAGTCGGTATCACTGC     GAATTGGTCGGTATCACTGC   \n",
       "19  kluyveromyces_marxianus  CACTTGCAGATGATCGCCAT                      NaN   \n",
       "20  kluyveromyces_marxianus  ATTGGTGGTGCCATTGGTAC                      NaN   \n",
       "21  kluyveromyces_marxianus  TGGTACTTACAACCAATTCA                      NaN   \n",
       "22  kluyveromyces_marxianus  TCTGCTACCAAGTGGATTGG                      NaN   \n",
       "23  kluyveromyces_marxianus  ACCAAGTGGATTGGTGGTCA                      NaN   \n",
       "24  kluyveromyces_marxianus  GACAGGAAGTTTGCCGACAT                      NaN   \n",
       "25  kluyveromyces_marxianus  GGTGAGAGATACAGAAAGGC                      NaN   \n",
       "26  kluyveromyces_marxianus        Uncovered gene                      NaN   \n",
       "27  kluyveromyces_marxianus  GTGAACATATCTCTTTGGAT                      NaN   \n",
       "28  kluyveromyces_marxianus  TCTCCATGTGATATGTGCAC                      NaN   \n",
       "\n",
       "    mismatch          strand        location gene_name      protein_id  \\\n",
       "0        2.0               F             854       NaN     XP_500614.1   \n",
       "1        NaN               F             224       NaN     XP_500977.1   \n",
       "2        NaN               F             227       NaN     XP_500977.1   \n",
       "3        1.0               F             236       NaN     XP_501094.1   \n",
       "4        2.0               F             242       NaN     XP_501094.1   \n",
       "5        2.0               F             254       NaN     XP_501094.1   \n",
       "6        1.0               F             254       NaN     XP_501094.1   \n",
       "7        NaN               F             635       NaN     XP_503263.1   \n",
       "8        NaN               F            2177       NaN     XP_503627.1   \n",
       "9        NaN              RC             204       NaN     XP_505753.2   \n",
       "10       NaN               F              76       NaN  XP_002489365.1   \n",
       "11       NaN               F             248       NaN  XP_002489665.1   \n",
       "12       NaN              RC            2775       NaN  XP_002490264.1   \n",
       "13       NaN               F             713       NaN  XP_002490740.1   \n",
       "14       NaN              RC             249       NaN  XP_002490927.1   \n",
       "15       NaN               F             287       NaN  XP_002492605.1   \n",
       "16       NaN               F             557       NaN  XP_002493751.1   \n",
       "17       NaN               F            1133       NaN  XP_002493751.1   \n",
       "18       2.0               F             884      CAN1  XP_022673643.1   \n",
       "19       NaN               F             275      GAP1  XP_022673914.1   \n",
       "20       NaN               F             293      GAP1  XP_022673914.1   \n",
       "21       NaN               F             352     MET17  XP_022674188.1   \n",
       "22       NaN               F             632     MET17  XP_022674188.1   \n",
       "23       NaN               F             638     MET17  XP_022674188.1   \n",
       "24       NaN               F             290      URA3  XP_022674242.1   \n",
       "25       NaN               F             761      URA3  XP_022674242.1   \n",
       "26       NaN  Uncovered gene  Uncovered gene      TRP1  XP_022677214.1   \n",
       "27       NaN              RC            2769      LYS2  XP_022678348.1   \n",
       "28       NaN               F             266      FCY1  XP_022678383.1   \n",
       "\n",
       "   ortho_prot_id ortho_gene_name  \n",
       "0    NP_010290.3            TRP1  \n",
       "1    NP_012965.3            GAP1  \n",
       "2    NP_012965.3            GAP1  \n",
       "3    NP_010851.1            CAN1  \n",
       "4    NP_010851.1            CAN1  \n",
       "5    NP_010851.1            CAN1  \n",
       "6    NP_010851.1            CAN1  \n",
       "7    NP_013406.1           MET17  \n",
       "8    NP_009673.1            LYS2  \n",
       "9    NP_015387.1            FCY1  \n",
       "10   NP_010290.3            TRP1  \n",
       "11   NP_012965.3            GAP1  \n",
       "12   NP_009673.1            LYS2  \n",
       "13   NP_010851.1            CAN1  \n",
       "14   NP_015387.1            FCY1  \n",
       "15   NP_010893.3            URA3  \n",
       "16   NP_013406.1           MET17  \n",
       "17   NP_013406.1           MET17  \n",
       "18   NP_010851.1            CAN1  \n",
       "19   NP_012965.3            GAP1  \n",
       "20   NP_012965.3            GAP1  \n",
       "21   NP_013406.1           MET17  \n",
       "22   NP_013406.1           MET17  \n",
       "23   NP_013406.1           MET17  \n",
       "24   NP_010893.3            URA3  \n",
       "25   NP_010893.3            URA3  \n",
       "26   NP_010290.3            TRP1  \n",
       "27   NP_009673.1            LYS2  \n",
       "28   NP_015387.1            FCY1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(f'data/output/test_{new_exp_name}_results.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allegro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
